### ISSUE TITLE
Unsupported hipster dot

### ISSUE THREAD

**Issue by themester** [cjbassi/gotop#18]
Gentoo.
x86_64.
st.
I do not use tmux xd.
go get.

![screenshot from 2018-04-10 13-08-55](https://user-images.githubusercontent.com/10146748/38554063-755001f6-3cc1-11e8-9d52-de52e82418b7.png)

I solved deleting braille chars from your library but I want you to know about this.

**Comment by themester** [comment]
@cjbassi Right now I am using rxvt-unicode and (I don't know why) does not support unicode braille characters xd

**Comment by AarynSmith** [comment]
I was having a similar issue with the graphs until I installed the Symbola font. Worth a shot if you don't have it installed already.

**Comment by yydcool** [comment]
I'm using Archlinux, solved by install package ttf-ubraille

**Comment by Yrp** [comment]
Is this problem solved, I have the same problem, my system is manjaro， I use terminator

**Comment by cjbassi** [comment]
ttf-ubraille didn't work for me on xterm or terminator unfortunately.

**Comment by Yrp** [comment]
By installing package ttf-ubraille and the Symbola font, and restart the terminator, the  issue has be solved.
I don't quite know which of these two packages works.

**Comment by NoMod-Programming** [comment]
Still not working on xterm :(

Also, the Debian repositories don't seem to have `ttf-ubraille`, so that's not an option for me.

**Comment by robotdna** [comment]
I wanted to +1 this issue with urxvt. I have installed ttf-ubraille, did not make any changes. Is there a way I can change the symbol manually with something else? Like a period or * ?

**Comment by cjbassi** [comment]
@robotdna Unfortunately not very easily atm. Gotop uses https://github.com/cjbassi/termui which uses https://github.com/cjbassi/drawille-go to draw the braille characters, and the logic is all hard coded there. I'm planning on fixing up https://github.com/gizak/termui and switching back to that, and once that happens then it should be easier to toggle the rendering character to something else.

**Comment by robotdna** [comment]
> @robotdna Unfortunately not very easily atm. Gotop uses https://github.com/cjbassi/termui which uses https://github.com/cjbassi/drawille-go to draw the braille characters, and the logic is all hard coded there. I'm planning on fixing up https://github.com/gizak/termui and switching back to that, and once that happens then it should be easier to toggle the rendering character to something else.

Okay, that makes sense. I'm not too broken up about it, it's not like the graph provides exact information haha. Thanks for looking into it.

**Comment by deckardstp** [comment]
![screenshot from 2018-12-30 22-38-21](https://user-images.githubusercontent.com/10142599/50551829-c19e3500-0c87-11e9-9730-f21408f90485.png)
I have a somehow similar topic. Since yesterday I only see graphs like this. Tried different terminals and it occurs in any. Standard is kitty. Is this related?

**Comment by cjbassi** [comment]
@deckardstp Similar but different issue for sure. Looks like the braille characters are rending fine, but it's also showing the empty braille cells in addition to the filled in line.

**Comment by cjbassi** [comment]
There was a reddit thread posted [here](https://www.reddit.com/r/archlinux/comments/apqnhr/ugly_graph_in_gotop/) where the OP was having this issue and he was able to fix it when using URxvt by installing the symbola font and loading it in the `.Xresources` file. Seems like installing symbola was working for some people but it looks like you also have to load it in `.Xresources` at least for URxvt. Hope that helps someone!

**Comment by robotdna** [comment]
> There was a reddit thread posted [here](https://www.reddit.com/r/archlinux/comments/apqnhr/ugly_graph_in_gotop/) where the OP was having this issue and he was able to fix it when using URxvt by installing the symbola font and loading it in the `.Xresources` file. Seems like installing symbola was working for some people but it looks like you also have to load it in `.Xresources` at least for URxvt. Hope that helps someone!

This solved it for me! To try to help those who are on a similar platformgr  as me with Arch+i3+urxvt:
1) grab ttf-symbola from AUR
2) fc-list | grep Symbola
3) Place the name in .Xresources as a font import. I did this:
URxvt*font: xft:Symbola:style=Regular:pixelsize=14
4) don't forget to rebuild with xrdb .Xresources

This worked, thanks!

**Comment by mksoc** [comment]
@duncanam Still not solved for me on a similar setup, but on Manjaro.

**Comment by woolenkitten** [comment]
I got the same issue with braille characters rendering after creating font configuration file (see [ fontconfig](https://wiki.archlinux.org/index.php/Font_configuration)). Without that file all works fine. [Here is how my fonts.conf looks](https://gist.github.com/woolenkitten/f908035c9f09985949180eb0e36843a4). Maybe I should add any rules for braille fonts? Can't find any information about that.

**Comment by dgrr** [comment]
I think it's just a matter of the font config you have. In Gentoo works fine using infinality.conf

**Comment by woolenkitten** [comment]
@dgrr I'm Ubuntu user. Also I'm using urxvt as my terminal emulator. Setting this line fixed rectangles in graphs
```
URxvt.font: xft:Fantasque Sans Mono:size=10:style=Bold, xft:symbola:size=11
```
I've just added `xft:symbola:size=11`, but seems like dots still not rendering correctly. Here is how it looks now: ![](https://i.imgur.com/iObHSHq.png). It looks like graph dots have wrong spacing.

**Comment by JPUnD** [comment]
I'm having the same issues, i'm running Manjaro with kitty tho, so the fixes don't apply to me (i've tried), anybody know a workaround?

**Comment by JPUnD** [comment]
> ![screenshot from 2018-12-30 22-38-21](https://user-images.githubusercontent.com/10142599/50551829-c19e3500-0c87-11e9-9730-f21408f90485.png)
> I have a somehow similar topic. Since yesterday I only see graphs like this. Tried different terminals and it occurs in any. Standard is kitty. Is this related?
> 
> The output of uname -a
> 4.19 & 4.20
> Terminal shell (like zsh or bash)
> zsh
> Terminal emulator (like iTerm or gnome terminal)
> kitty
> Using tmux?
> no

Did you find a solution? The issue still persist on my setup.

**Comment by deckardstp** [comment]
> Did you find a solution? The issue still persist on my setup.

My original installation was deleted in the meantime. I installed a version from AUR which didn't show the issue as far as I can remember. Also I switched to tilix due to the better window tiling. My new manjaro installation doesn't show this behavior. Sorry, that I can't help you any better.

**Comment by duncanam** [comment]
@JPUnD Do you see any fonts I have in my config that you don't have enabled for your terminal? I got mine working, at least with the ones I have [**here**.](https://github.com/duncanam/dotfiles/blob/master/.Xresources)

**Comment by JPUnD** [comment]
> @JPUnD Do you see any fonts I have in my config that you don't have enabled for your terminal? I got mine working, at least with the ones I have [**here**.](https://github.com/duncanam/dotfiles/blob/master/.Xresources)

I've just realized that kitty can only use monospaced fonts, due to the way it renders text, and although i have those installed in my system, i cannot use them in kitty if they are not monospaced (it doesn't even show in the kitty command to list all available fonts).

Sadly none of the fonts you have displayed, and i've installed on my system, are monospaced.

Should i take this issue to kitty?

**Comment by gkeep** [comment]
@JPUnD There is a [fix](https://github.com/kovidgoyal/kitty/issues/1623#issuecomment-493767496) for kitty. As suggested, you need to install symbola font (`ttf-symbola` in AUR), then add `symbol_map U+2800-U+28FF Symbola` to your `kitty.conf`.

![image](https://user-images.githubusercontent.com/20600053/66274361-89ff9180-e886-11e9-93c7-68ae3ede5781.png)

**Comment by JPUnD** [comment]
@gkeep Damn, how did i miss that? Anyways, it does in fact work, since it replaces the character that was giving me trouble with, thanks!

**Comment by omarkamalakis** [comment]
Hey I had a similar problem on my machine and I fixed it by removing `gnu-free-fonts`. Apparently, free-font-mono was overriding the other fonts with its own type of braille (the one with the empty circles). I deleted it and installed another braille font (Dejavu Sans works nicely, as does ubraille) and it worked. Use `gucharmap` to see where your problematic fonts are.

**Comment by arbaes** [comment]
> Hey I had a similar problem on my machine and I fixed it by removing `gnu-free-fonts`. Apparently, free-font-mono was overriding the other fonts with its own type of braille (the one with the empty circles).

That's the issue I had, I would have preferred to find where to config the fallback font instead of uninstalling the one set. But it did the trick for me so thanks.

**Comment by deckardstp** [comment]
> @JPUnD There is a [fix](https://github.com/kovidgoyal/kitty/issues/1623#issuecomment-493767496) for kitty. As suggested, you need to install symbola font (`ttf-symbola` in AUR), then add `symbol_map U+2800-U+28FF Symbola` to your `kitty.conf`.
> 
> ![image](https://user-images.githubusercontent.com/20600053/66274361-89ff9180-e886-11e9-93c7-68ae3ede5781.png)

thanks! worked great for me on manjaro. `ttf-symbola` is now outdated and the successor is `ttf-symbola-free`.

**Comment by markuspeloquin** [comment]
tl;dr I have a real solution for anybody who needs it :). Get rid of gnu-free-mono-fonts (fedora) fonts-freefont-ttf (I think? debian).

---

So I found a solution-ish here, which I modified to be easier and specific: https://unix.stackexchange.com/questions/247108/how-to-find-out-which-unicode-codepoints-are-defined-in-a-ttf-file

```python
#!/usr/bin/env python3

import os
import sys
from fontTools.ttLib import TTFont

def find(root: str) -> None:
    print(f'Looking for U+2800 in {root}')
    for dirpath, dirnames, filenames in os.walk(root):
        for name in filenames:
            base, ext = os.path.splitext(name)
            if ext.lower() not in ('.ttf', '.otf'): continue
            pathname = os.path.join(dirpath, name)
            try:
                font = TTFont(pathname)
                for cmap in font['cmap'].tables:
                    if cmap.isUnicode():
                        if 0x2800 in cmap.cmap:
                            print(f'Found in {pathname}')
                            break
            except Exception as e:
                print(f'Failed to read {pathname}: {e}')

for dir in [os.path.join(os.environ['HOME'], '.fonts'), '/usr/share/fonts']:
    if os.path.isdir(dir):
        find(dir)
```

`pip install fonttools` and run it. It showed which fonts *actually* provide Braille symbols, because there's only a few. I see UBraille and Symbola (a couple attempts of mine to fix this), unifont, FreeMono, and DejaVu Sans/Serif fonts. That grubby GNU FreeMono font was the problem. Get rid of it. In Fedora, it was gnu-free-mono-fonts, and nothing required it.

Then restarting the process fixed it.

**Comment by dosssman** [comment]
Greetings.

Maybe a little bit too late, but the following steps happened to solve the issue with gotop's CPU lines not showing  up correctly, as described in the first messages.

The OS is Arch Linux, and the terminal is `termite`.

1. Installing the Braille font package (`ttf-ubraille` on the AUR. Will likely differ for other distributions).
2. Regenerate the font cache with `fc-cache -fv`, or equivalent.
2. Edit the `~/.config/fontconfig/fonts.conf` according to the font used by the terminal. The goal is to set the Braille font as a fallback option, in case the characte gotop needs to display is not found in the default font. In my case, termite was set to use the Monospace font, so the file looks as follows:
```
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE fontconfig SYSTEM "fonts.dtd">
 <!-- -->
 <fontconfig>
   <!-- Generic name aliasing -->
   <alias>
     <family>monospace</family>
     <prefer>
       <family>Liberation Mono</family>
       <family>Braille</family>
    </prefer>
  </alias>
</fontconfig>
```
After this, gotop's line are displayed properly.

**Comment by xxxserxxx** [comment]
Thanks @dosssman. gotop (active development) lives over at [xxxserxxx/gotop](https://github.com/xxxserxxx/gotop) now.
